### Binary Classification
For binary classification with Random Forest Regressor

In [2]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns
%matplotlib inline 
from sklearn import preprocessing
from math import sqrt
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
pd.set_option('display.max_colwidth', -1)

In [5]:
#Load red wine data
data = pd.read_csv('../machine/heart.csv')
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0  63   1    3   145       233   1    0        150      0      2.3      0       
1  37   1    2   130       250   0    1        187      0      3.5      0       
2  41   0    1   130       204   0    0        172      0      1.4      2       
3  56   1    1   120       236   0    1        178      0      0.8      2       
4  57   0    0   120       354   0    1        163      1      0.6      2       

   ca  thal  target  
0  0   1     1       
1  0   2     1       
2  0   2     1       
3  0   2     1       
4  0   2     1

In [6]:
data.shape

(303, 14)

We have 303 samples and 14 features, including our target feature. We can easily print some summary statistics.

In [7]:
#Summary statisticsPython
data.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean   54.366337   0.683168    0.966997    131.623762  246.264026  0.148515     
std    9.082101    0.466011    1.032052    17.538143   51.830751   0.356198     
min    29.000000   0.000000    0.000000    94.000000   126.000000  0.000000     
25%    47.500000   0.000000    0.000000    120.000000  211.000000  0.000000     
50%    55.000000   1.000000    1.000000    130.000000  240.000000  0.000000     
75%    61.000000   1.000000    2.000000    140.000000  274.500000  0.000000     
max    77.000000   1.000000    3.000000    200.000000  564.000000  1.000000     

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean   0.528053    149.646865  0.326733    1.039604    1.399340    0.729373     
std    0.525860    22.905161   0.469794    1.161075    0.616226    1.022606     
min    0.000000    71.000000   0.000000    0.000000    0.000000    0.000000     
25%    0.000000    133.500000  0.000000    0.000000    1.000000    0.000000     
50%    1.000000    153.000000  0.000000    0.800000    1.000000    0.000000     
75%    1.000000    166.000000  1.000000    1.600000    2.000000    1.000000     
max    2.000000    202.000000  1.000000    6.200000    2.000000    4.000000     

             thal      target  
count  303.000000  303.000000  
mean   2.313531    0.544554    
std    0.612277    0.498835    
min    0.000000    0.000000    
25%    2.000000    0.000000    
50%    2.000000    1.000000    
75%    3.000000    1.000000    
max    3.000000    1.000000

First, let's separate our target (y) features from our input (X) features:

In [9]:
# Separate target from training features P
y = data.target
X = data.drop('target', axis=1)

This allows us to take advantage of Scikit-Learn's useful train_test_split function:

In [10]:
#Split data into train and test setsPython

X_train, X_test, y_train, y_test = train_test_split(X, y, 
                                                    test_size=0.2, 
                                                    random_state=123, 
                                                    stratify=y)

As you can see, we'll set aside 20% of the data as a test set for evaluating our model. We also set an arbitrary "random state" (a.k.a. seed) so that we can reproduce our results.

Finally, it's good practice to stratify your sample by the target variable. This will ensure your training set looks similar to your test set, making your evaluation metrics more reliable.

#### Data preprocessing 

Set up the cross-validation pipeline

In [11]:
# Pipeline with preprocessing and model
pipeline = make_pipeline(preprocessing.StandardScaler(), 
                         RandomForestRegressor(n_estimators=100))

This is exactly what it looks like: a modeling pipeline that first transforms the data using StandardScaler() and then fits a model using a random forest regressor.

#### Declare hyperparameters to tune

Now, let's declare the hyperparameters we want to tune through cross-validation.

In [12]:
# Declare hyperparameters to tune
hyperparameters = { 'randomforestregressor__max_features' : ['auto', 'sqrt', 'log2'],
                  'randomforestregressor__max_depth': [None, 5, 3, 1]}

In [14]:
#Sklearn cross-validation with pipeline
clf = GridSearchCV(pipeline, hyperparameters, cv=10)
 
# Fit and tune model
clf.fit(X_train, y_train)

GridSearchCV(cv=10, error_score=nan,
             estimator=Pipeline(memory=None,
                                steps=[('standardscaler',
                                        StandardScaler(copy=True,
                                                       with_mean=True,
                                                       with_std=True)),
                                       ('randomforestregressor',
                                        RandomForestRegressor(bootstrap=True,
                                                              ccp_alpha=0.0,
                                                              criterion='mse',
                                                              max_depth=None,
                                                              max_features='auto',
                                                              max_leaf_nodes=None,
                                                              max_samples=None,
                            

#### Evaluate model pipeline on test data

In [15]:
#Predict a new set of dataPython
y_pred = clf.predict(X_test)

In [35]:
from sklearn.metrics import classification_report, confusion_matrix
print(accuracy_score(y_test, y_pred.round()))
print(confusion_matrix(y_test,y_pred.round()))
print(classification_report(y_test,y_pred.round()))


0.8688524590163934
[[22  6]
 [ 2 31]]
              precision    recall  f1-score   support

           0       0.92      0.79      0.85        28
           1       0.84      0.94      0.89        33

    accuracy                           0.87        61
   macro avg       0.88      0.86      0.87        61
weighted avg       0.87      0.87      0.87        61



You can see from the confusion matrix that our Random Forest Regressor only misclassified two out of the 61. Also, the f1 score of 0.87 .